Bearded.Utilities/Monads/Result.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Bearded.Utilities.Monads
+{
+    public readonly struct Result<TResult, TError> : IEquatable<Result<TResult, TError>>
+    {
+        private readonly bool isSuccess;
+        private readonly TResult result;
+        private readonly TError error;
+
+        private Result(bool isSuccess, TResult result, TError error)
+        {
+            this.isSuccess = isSuccess;
+            this.result = result;
+            this.error = error;
+        }
+
+        public TResult ResultOrDefault(TResult @default) => isSuccess ? result : @default;
+
+        public TResult ResultOrDefault(Func<TResult> defaultProvider) => isSuccess ? result : defaultProvider();
+
+        public TResult ResultOrThrow(Func<TError, Exception> exceptionProvider) =>
+            isSuccess ? result : throw exceptionProvider(error);
+
+        public Result<TOut, TError> Select<TOut>(Func<TResult, TOut> selector) =>
+            isSuccess ? (Result<TOut, TError>) Result.Success(selector(result)) : Result.Failure(error);
```

Comment:
Why not (and same below)
```suggestion
            isSuccess ? (Result<TOut, TError>) Result.Success(selector(result)) : this;
```

Reply:
Because the success type isn't the same, and you can't cast `Result<TResult, TError>` to `Result<TOut, TError>`, unless you case it back to a `Failure<TError>` first.